Sources/_SwiftSyntaxTestSupport/FixItApplier.swift

```diff
@@ -0,0 +1,107 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftDiagnostics
+import SwiftSyntax
+
+public class FixItApplier {
+  fileprivate struct Edit {
+    let startUtf8Offset: Int
+    let endUtf8Offset: Int
+    let replacement: String
+  }
+
+  private let changes: [FixIt.Change]
+  private let tree: any SyntaxProtocol
+
+  private lazy var locationConverter: SourceLocationConverter = {
+    return SourceLocationConverter(fileName: "", tree: tree)
+  }()
+
+  init(diagnostics: [Diagnostic], withMessages messages: [String]?, tree: any SyntaxProtocol) {
+    let messages = messages ?? diagnostics.compactMap { $0.fixIts.first?.message.message }
```

Comment:
Am I understanding this correctly? When `messages`` is nil, we want to include all the Fix-Its from the `diagnostics``, right? However, doesn't this code `diagnostics.compactMap { $0.fixIts.first?.message.message }`` only add the message from the first Fix-It of each diagnostic?

If my understanding is correct, perhaps we should have something like this instead:

```swift
self.changes =
      diagnostics
      .flatMap(\.fixIts)
      .filter { messages?.isEmpty() == true ? messages?.contains($0.message.message) : false } // Is nil representing the same state as empty array? If not instead of `.isEmpty()` we can use `== nil ? false : messages?.contains($0.message.message)`
      .flatMap(\.changes) 
```
instead.

Reply:
The reason is because some diagnostics give us two Fix-Its.
So we default to the first one.

Reply:
I'm thinking... we might need to revisit this based on what we've stated in PR #2021. According to the documentation for the `assert` method there, if `applyFixIts` is set to `nil`, it should apply all Fix-Its from the diagnostics. If we continue to only map the first Fix-It, we might end up with a discrepancy between the code behavior and the documentation, which could be confusing for end-users. 

I'm not sure if you are going to address it in the #2021 PR so I've added this comment here 🫶

Reply:
We should document it yes.
The reason is because when we have a Fix-It that offer two solutions, what one should we apply?
We cannot apply both of them.

You can take a loot at this example etc
https://github.com/apple/swift-syntax/blob/7c6d4f729f37f5e5a206b39ffa7bc1fac9479c24/Tests/SwiftParserTest/translated/RecoveryTests.swift#L993-L1017

Reply:
Ah, I get it now! Thanks for clarifying that the `fixIts` collection in `Diagnostic` is designed to offer different solutions for resolving the same diagnostic issue. I was under the impression that it's a collection where each Fix-It is a part of a multi-step solution for the problem at hand. For example, one changing a particular node and another modifies something else to fully resolve the issue. Now it all makes sense! Appreciate the explanation. 🙈